### PR TITLE
Added dynamic room functionality

### PR DESCRIPTION
**Pull request by Anjali-Rightpoint**
Added functionality to display conference rooms location wise.

### CODE REVIEW

File: RoomFinder/Room List/RoomListViewModel.swift [hunk 1]
```diff
@@ -8,17 +8,18 @@
 
 import SwiftUI
 import Combine
+import Services
 
 class RoomListViewModel: ObservableObject {
     let locationName: String
+    var roomsArray = [RoomsListValue]()
```
Comment (pauluhn) [inline RoomFinder/Room List/RoomListViewModel.swift:15]: #nth drop `Array`

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 2]
```diff
@@ -22,21 +23,29 @@ class LocationListViewModel: ObservableObject {
...
-        RoomManager().fetchListOfRooms { (roomsData) in
-            for room in roomsData {
-                if let city = room.address?.city {
-                    if !(self.data.contains(city)) {
-                        self.data.append(city)
-                    }
-                }
-            }
-            self.updateData()
+        if (roomsListBasedOnLocations == [:]){
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:26]: #req use swift-y way to check for empty state

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 3]
```diff
@@ -22,21 +23,29 @@ class LocationListViewModel: ObservableObject {
...
+            RoomManager().fetchListOfRooms { (roomsData) in
+                       for room in roomsData {
+                           if let city = room.address?.city {
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:29]: #pp use guard and combine with second `if`

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 4]
```diff
@@ -22,21 +23,29 @@ class LocationListViewModel: ObservableObject {
...
+                               if !(self.data.contains(city)) {
+                                   if((self.roomsListBasedOnLocations[city]) != nil){
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:31]: #nth make swift-y

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 5]
```diff
@@ -22,21 +23,29 @@ class LocationListViewModel: ObservableObject {
...
         }
+       
     }
     
     
+    
     func updateData() {
-        self.dataSource = data.map({ name in
+        self.dataSource = roomsListBasedOnLocations.keys.map({ name in
             return LocationRowViewModel(name: name)
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:49]: #req use identifier

File: RoomFinder/Room List/RoomListViewModel.swift [hunk 6]
```diff
@@ -8,17 +8,18 @@
...
     var dataSource: [RoomRowViewModel] = []
     
-    let mockData: [String] = ["The Den", "Oasis", "Matrix", "The Garage"]
     
-    init(locationName: String) {
+    init(locationName: String, roomsArray: [RoomsListValue]) {
         self.locationName = locationName
-        self.dataSource = mockData.map({ name in
-            return RoomRowViewModel(name: name)
+        self.dataSource = roomsArray.map({ room in
```
Comment (pauluhn) [inline RoomFinder/Room List/RoomListViewModel.swift:21]: #pp drop parentheses

File: RoomFinder/Room List/RoomRowViewModel.swift [hunk 7]
```diff
@@ -16,9 +16,13 @@ struct RoomRowViewModel: Identifiable {
     }
     
     let name: String
+    let capacity: Int
+    let emailAddress: String
     
-    init(name: String) {
+    init(name: String,capacity: Int,emailAddress: String) {
```
Comment (pauluhn) [inline RoomFinder/Room List/RoomRowViewModel.swift:22]: #nth default values?

File: RoomFinder/Location List/LocationListViewModel.swift (hunk 5 above)
Comment (Anjali-Rightpoint) [reply RoomFinder/Location List/LocationListViewModel.swift:46]: Need some more clarification on this

Comment (pauluhn) [reply RoomFinder/Location List/LocationListViewModel.swift:46]: to me it looks like you are using `name` to determine the model to use in `LocationRowViewModel`. it should be an `id` instead.

Comment (pauluhn) [reply RoomFinder/Location List/LocationListViewModel.swift:46]: or the actual model

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 8]
```diff
@@ -22,31 +23,34 @@ class LocationListViewModel: ObservableObject {
     }
     
     func fetchLocationsData() {
-        RoomManager().fetchListOfRooms { (roomsData) in
-            for room in roomsData {
-                if let city = room.address?.city {
-                    if !(self.data.contains(city)) {
-                        self.data.append(city)
+        if (roomsListBasedOnLocations.isEmpty){
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:26]: #req when there is no `else` best to use `guard`

File: RoomFinder/Room List/RoomListViewModel.swift [hunk 9]
```diff
@@ -8,18 +8,18 @@
...
-    let mockData: [String] = ["The Den", "Oasis", "Matrix", "The Garage"]
     
-    init(locationName: String) {
+    init(locationName: String, rooms: [RoomsListValue]) {
         self.locationName = locationName
-        self.dataSource = mockData.map({ name in
-            return RoomRowViewModel(name: name)
-        })
+        self.dataSource = rooms.map{ room in
+            return RoomRowViewModel(name: room.displayName!,capacity: room.capacity ?? 0,emailAddress: room.emailAddress ?? "")
```
Comment (pauluhn) [inline RoomFinder/Room List/RoomListViewModel.swift:21]: #req you've added default values so you can remove the nil-coalescing

File: RoomFinder/Location List/LocationListViewModel.swift (hunk 5 above)
Comment (Anjali-Rightpoint) [reply RoomFinder/Location List/LocationListViewModel.swift:47]: roomsListBasedOnLocations is a dictionary which has locations as keys and the value is the roomArray for that particular location.
So here we are creating a row model(LocationRowViewModel) using a locationName and the locationName will be used as id also in that model.

File: RoomFinder/Room List/RoomListViewModel.swift (hunk 9 above)
Comment (Anjali-Rightpoint) [reply RoomFinder/Room List/RoomListViewModel.swift:21]: As per my understanding, Default value vs nil-coalescing is as below

- Default value gets used when we don't pass the param from the caller.

For Ex:
Caller :
    foo()

Function definition: -
    func foo(param: String = ""){
    }

- nil-coalescing is required when we have function paramater as non-optional and the value which we are passing from caller is an optional

For Ex:
var temp: String?

Approach 1: (As of now we are using this approach)

Caller
    foo(param: temp ?? "")

Function definition:

    func foo(param: String = ""){
    }

Approach 2:

Caller:
    foo(param: temp)

Function definition: 
    func foo(param: String? = ""){
    // in this case, we need to either use nil-coalescing or force unwrapping while using param
    }

Please let me know if my understanding is correct and which approach we should follow.

File: RoomFinder/Location List/LocationListViewModel.swift [hunk 10]
```diff
@@ -22,31 +23,35 @@ class LocationListViewModel: ObservableObject {
     }
     
     func fetchLocationsData() {
+       
+        guard roomsListBasedOnLocations.isEmpty else{
```
Comment (pauluhn) [inline RoomFinder/Location List/LocationListViewModel.swift:27]: #? are we able to force-refresh the locations data? I agree that it probably won't change much.

File: RoomFinder/Location List/LocationListViewModel.swift (hunk 5 above)
Comment (pauluhn) [reply RoomFinder/Location List/LocationListViewModel.swift:47]: its just semantics. I'd prefer it were clear that it was an identifier even if it is backed by a string. but I'll change it to `#nth` instead

File: RoomFinder/Room List/RoomListViewModel.swift (hunk 9 above)
Comment (pauluhn) [reply RoomFinder/Room List/RoomListViewModel.swift:21]: you are correct that we need to accept `optional` and then unwrap with default values. it just makes the call site easier to read. however, we can change this to `#nth` as well.

**Pull request merged**